### ISSUE TITLE
Blazor @bind:after not working on .NET 7 RTM release

### ISSUE THREAD

**Issue by iphdav** [dotnet/aspnetcore#44957]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Cleanly installed VS 17.4, which installed .NET 7:
dotnet --list-sdks
7.0.100 [C:\Program Files\dotnet\sdk]

Attempting to run example component:
```
@page "/"

<input @bind="searchText" @bind:after="PerformSearch" />

@code {
    string searchText;

    async Task PerformSearch()
    {
        // ... do something asynchronously with 'searchText' ...
    }
}
```

Project file:
```
<Project Sdk="Microsoft.NET.Sdk.Web">

  <PropertyGroup>
    <TargetFramework>net7.0</TargetFramework>
    <Nullable>enable</Nullable>
    <ImplicitUsings>enable</ImplicitUsings>
  </PropertyGroup>

</Project>
```

dotnet build
MSBuild version 17.4.0+18d5aef85 for .NET

Get following error:
C:\Users\dtaylor\source\repos\EmptyBlazorApp1\Microsoft.NET.Sdk.Razor.SourceGenerators\Microsoft.NET.Sdk.Razor.SourceGe
nerators.RazorSourceGenerator\Pages_Index_razor.g.cs(56,132): error CS1503: Argument 3: cannot convert from 'Microsoft.
AspNetCore.Components.EventCallback<string>' to 'System.Action<string?>' [C:\Users\dtaylor\source\repos\EmptyBlazorApp1
\EmptyBlazorApp1.csproj]

### Expected Behavior

Expected behaviour is sample included at URL below should work as documented in the blog entry:
https://devblogs.microsoft.com/dotnet/asp-net-core-updates-in-dotnet-7-preview-7/#:~:text=In.NET%207%20you%20can%20now%20easily%20run%20async,%2F%2F...%20do%20something%20asynchronously%20with%20%27searchText%27...%20%7D%20%7D

### Steps To Reproduce

Just use:
dotnet new blazorserver-empty

Then paste this code into the Pages/Index.razor file:

```
@page "/"

<input @bind="searchText" @bind:after="PerformSearch" />

@code {
    string searchText;

    async Task PerformSearch()
    {
        // ... do something asynchronously with 'searchText' ...
    }
}
```

### .NET Version

7.0.100

### Anything else?

.NET SDK:
 Version:   7.0.100
 Commit:    e12b7af219

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19042
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\7.0.100\

Host:
  Version:      7.0.0
  Architecture: x64
  Commit:       d099f075e4

.NET SDKs installed:
  7.0.100 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 6.0.11 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  x86   [C:\Program Files (x86)\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x86\InstallLocation]

Environment variables:
  Not set

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download

**Comment by iphdav** [comment]
I just read the previous issue https://github.com/dotnet/aspnetcore/issues/44955 and expect my issue is related. 

This is one of the most touted features of .NET 7 / Blazor (talked about at .NET Conf) - so if this is broken it needs to be an immediately priority to get a patch out for the RTM release.

**Comment by javiercn** [comment]
@iphdav thanks for contacting us.

This is a known issue. https://github.com/dotnet/core/blob/main/release-notes/7.0/known-issues.md#70-bind-get-set-after-cant-be-used-in-70-blazor-applications

We had an issue with our dependency flow and one commit did not make it into RTM. A fix will be available in the first patch release.

**Comment by Alerinos** [comment]
@javiercn 
Can I have a link please?
Unfortunately, this version is not here.
https://dotnet.microsoft.com/en-us/download/dotnet/7.0

**Comment by javiercn** [comment]
https://github.com/dotnet/installer 

Release/7.0.2xx
(7.0.x Runtime)

**Comment by javiercn** [comment]
@mbobka run `dotnet --info` to make sure you are actually using the `7.0.200-preview.22558.7` SDK (or use a global.json). (`dotnet new globaljson --sdk-version 7.0.200-preview.22558.7`).

Clean the bin and obj folders to avoid any issue caused by caching. Run `dotnet build` from the command line.

I believe Visual Studio might still give you errors, but the command line should be fine.

**Comment by mbobka** [comment]
@javiercn 
`Welcome to .NET 7.0!
---------------------
SDK Version: 7.0.200-preview.22558.7

Telemetry
---------
The .NET tools collect usage data in order to help us improve your experience. It is collected by Microsoft and shared with the community. You can opt-out of telemetry by setting the DOTNET_CLI_TELEMETRY_OPTOUT environment variable to '1' or 'true' using your favorite shell.

Read more about .NET CLI Tools telemetry: https://aka.ms/dotnet-cli-telemetry

----------------
Installed an ASP.NET Core HTTPS development certificate.
To trust the certificate run 'dotnet dev-certs https --trust' (Windows and macOS only).
Learn about HTTPS: https://aka.ms/dotnet-https
----------------
Write your first app: https://aka.ms/dotnet-hello-world
Find out what's new: https://aka.ms/dotnet-whats-new
Explore documentation: https://aka.ms/dotnet-docs
Report issues and find source on GitHub: https://github.com/dotnet/core
Use 'dotnet --help' to see available commands or visit: https://aka.ms/dotnet-cli
--------------------------------------------------------------------------------------
MSBuild version 17.5.0-preview-22506-03+20ce296d6 for .NET`

**Comment by mbobka** [comment]
```
.NET SDK:
 Version:   7.0.200-preview.22558.7
 Commit:    6370b56ac2

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.22000
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\7.0.200-preview.22558.7\

Host:
  Version:      7.0.0
  Architecture: x64
  Commit:       d099f075e4

.NET SDKs installed:
  7.0.100 [C:\Program Files\dotnet\sdk]
  7.0.101 [C:\Program Files\dotnet\sdk]
  7.0.200-preview.22558.7 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.0-rtm.22504.29 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.0-rtm.22512.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.0-rtm.22505.15 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.0-rtm.22511.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 6.0.11 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.0-rtm.22478.5 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  x86   [C:\Program Files (x86)\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x86\InstallLocation]

Environment variables:
  Not set

global.json file:
  C:\xp\?????????? ????????\global.json

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download
  ```

**Comment by Alerinos** [comment]
@mbobka 
```razor
<input @bind:get="Value" @bind:set="ValueChanged" />

@code {
    [Parameter]
    public string? Value { get; set; }

    [Parameter]
    public EventCallback<string> ValueChanged { get; set; }
}
```

https://learn.microsoft.com/en-us/aspnet/core/blazor/components/data-binding?view=aspnetcore-7.0#:~:text=Components%20support%20two%2Dway%20data%20binding%20by%20defining%20a%20pair%20of%20parameters%3A

Check that you have used the component correctly. This is an example from the documentation.

In my case, the update helped and everything works fine.

**Comment by mbobka** [comment]
@Alerinos 
I'm trying to compile this code:
```
@page "/"

<input @bind="searchText" @bind:after="PerformSearch" />

@code {
    string searchText;

    async Task PerformSearch()
    {
        // ... do something asynchronously with 'searchText' ...
    }
}
```

**Comment by Alerinos** [comment]
@mbobka 

Error occurs during compilation or also shows in visual studio?

If this is a problem with compilation, try to clean the solution and then compile the program again.
If this is a visual studio problem, I recommend restarting the program, clearing the cache and even restarting the device. I had a similar case once with preview .net 6

**Comment by Alerinos** [comment]
@mbobka 
https://github.com/Alerinos/Virtualize-bug/blob/main/Pages/BindTest.razor

I just checked, download the repository and check it out. If it doesn't work, there is probably a problem with your device.
As a last resort, you can reinstall visual studio and SDK.

**Comment by mbobka** [comment]
The problem occurs only with Blazor WebAssembly apps, that refences release packages:
```
    <PackageReference Include="Microsoft.AspNetCore.Components.WebAssembly" Version="7.0.0" />
    <PackageReference Include="Microsoft.AspNetCore.Components.WebAssembly.DevServer" Version="7.0.0" PrivateAssets="all" />
```
If I change release packages on previous rc.2 versions then all goes well:
```
    <PackageReference Include="Microsoft.AspNetCore.Components.WebAssembly" Version="7.0.0-rc.2.22476.2" />
    <PackageReference Include="Microsoft.AspNetCore.Components.WebAssembly.DevServer" Version="7.0.0-rc.2.22476.2" PrivateAssets="all" />
```

**Comment by javiercn** [comment]
To be very clear, using a new SDK won't work in VS. Only from the command line, since the compiler ships separately to VS and to the SDKs. Its already on the SDKs and it will soon (if not already) be on a newer VS version.

**Comment by almightyju** [comment]
> @mbobka I would avoid that.
> 
> We changed the compiler APIs for RTM, you are essentially relying on old APIs that produce additional renders. The compiler was updated at the same time, but missed the RTM train. It is best to move to a newer SDK version rather than backwards.

I'm 100% using the latest build of dotnet (7.0.200-preview.22558.7) and building from the command line; the RC packages is the only way to get this to work. 

Do you know if there's likely to be bugs introduced by developing against the RC packages for now, or is it just performance improvements?

**Comment by CrahunGit** [comment]
I’ve tried new ask and command line and it works fine like @javiercn said. And it fails on vs right now. Waiting for the new vs update

**Comment by almightyju** [comment]
> I’ve tried new sdk and command line works fine like @javiercn said. And it fails on vs right now. Waiting for the new vs update

Are you using the webassembly packages?

**Comment by mbobka** [comment]
> To be very clear, using a new SDK won't work in VS. Only from the command line, since the compiler ships separately to VS and to the SDKs. Its already on the SDKs and it will soon (if not already) be on a newer VS version.

It works in VS 17.5.0 preview 1.0. But SDK I installed separately.

**Comment by CrahunGit** [comment]
> > I’ve tried new sdk and command line works fine like @javiercn said. And it fails on vs right now. Waiting for the new vs update
> 
> Are you using the webassembly packages?

Of course. I’m using webassembly.

**Comment by almightyju** [comment]
> Of course. I’m using webassembly.

Then I have no idea how we are getting different results 😞 I've got the gobal.json file in place and even made a whole new blazor app and only build from the command line and it still fails.

**Comment by ladeak** [comment]
Could you share when we may expect the update non-prerelease 7.0.x update released?

**Comment by GerardSmit** [comment]
> > Of course. I’m using webassembly.
> 
> Then I have no idea how we are getting different results 😞 I've got the gobal.json file in place and even made a whole new blazor app and only build from the command line and it still fails.

Same here.

After installing .NET 7.0.2, Visual Studio 2022 Preview and creating a `global.json` I still have this problem in Blazor Server-side.

<details>
<summary>Commands</summary>
<p>

```
$ dotnet build
MSBuild version 17.5.0-preview-22567-05+7037436dc for .NET

...

C:\...\Microsoft.NET.Sdk.Razor.SourceGenerators\Microsoft.NET.Sdk.Razor.SourceGenerators.RazorSourceGenerator\Shared_SiteSelector_razor.g.cs(248,137): error CS1503: Argument 3: cannot convert from 'Microsoft.AspNetCore.Components.EventCallback<string
>' to 'System.Action<string?>' [C:\...\....csproj]
```

```

$ dotnet --info
.NET SDK:
 Version:   7.0.200
 Commit:    27f0a7fa5a

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.22621
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\7.0.200\

Host:
  Version:      7.0.0
  Architecture: x64
  Commit:       d099f075e4

.NET SDKs installed:
  6.0.403 [C:\Program Files\dotnet\sdk]
  7.0.100 [C:\Program Files\dotnet\sdk]
  7.0.200 [C:\Program Files\dotnet\sdk]
```

</p>
</details>

**Comment by florianwachs** [comment]
Same here, installed 7.0.2 daily, cleaned solution in new blazor server app. build on terminal with dotnet build -> same EventCallback error.

**Comment by florianwachs** [comment]
@javiercn Could you please confirm that this is fixed for Blazor Server too? I wasn't able to reproduce that it is fixed in the daily 7.0.2 version, but for me it would be enough to know that it is fixed in the next service release. Thanks a lot!

**Comment by wolfgangschneider** [comment]
FYI - still not working Blazor-Server 7.0.2 SDK

![image](https://user-images.githubusercontent.com/15703770/202838727-a1816b8f-2eb0-4f26-b2fd-6a63e89d9853.png)

**Comment by eyalalonn** [comment]
I'm getting this error too when using `@bind:get="@Value" @bind:set="@ValueChanged"` and I'm using the latest SDK `7.0.200` as suggested above. Is there any ETA?

**Comment by eyalalonn** [comment]
@javiercn Is there any workaround for this?

**Comment by javiercn** [comment]
@eyalalonn Unfortunately, this is a mismatch between the code generated by the compiler and the runtime APIs available. 

We are working to get the compiler updated in the SDK, but a fix will not be available officially in until 7.0.102.
We will however update this thread when we have a nightly build of the SDK available with the fix to unblock folks.

In a previous post I incorrectly pointed out to a different version of the SDK, that seemed to fix the issue, but the issue was that it contained older bits than the RTM version and that's what made it work.

**Comment by eyalalonn** [comment]
@javiercn Thank you.

**Comment by fdonnet** [comment]
just installed 17.4.2 Visual Studio (not working). Waiting this feature for a long time (ability to trigger some JS after binding compared to explicitly call the function from the parent compo OnAfterRenderAsync func)... 
Do you have any date for the release on the normal branch (not preview). If it is too long, I will go with my good old dirty way that triggers bad dependencies between my compo (don't forget to run that after...).

**Comment by iphdav** [comment]
@javiercn Isn't 7.0.101 the next version out, so does this mean the fix isn't even going to make it into the next update to .NET 7 SDK?

**Comment by MisinformedDNA** [comment]
Since this is still broken, how would I do this if I were still using .NET 6? There must be some workaround we can use.

**Comment by fdonnet** [comment]
@MisinformedDNA it will not work depending on what you are doing and if you use JS interop. In some case you need to trigger something in JS after your data have been bind... that's why this bind:after will be so useful. On my side, for the moment, I trigger my JS from my parent compo when I m sure the data are present... but it' is not clean and create bad dependencies.

**Comment by gavilanch** [comment]
.NET 7.0.1 was released. And even after updating that and Visual Studio to the latest Version 17.4.3, I still get the following error in a newly created .NET 7 project: 

`CS1503	Argument 1: cannot convert from 'Microsoft.AspNetCore.Components.EventCallback<string>' to 'System.Func<string, System.Threading.Tasks.Task>'`

My code is the following:

```
@page "/"

<input @bind:get="Value" @bind:set="ValueChanged" />

@code {
    [Parameter]
    public string? Value { get; set; }

    [Parameter]
    public EventCallback<string> ValueChanged { get; set; }
}
```

According to the docs, this should work since I'm using 7.0.1.

I get the same situation even when creating the project using the dotnet CLI and vs code.

Am I missing something?

**Comment by mikes-gh** [comment]
I looked at the 7.0.1 release notes and theeeres no mention of any fixes to the bind attributes.
Can someone form MS please clarify

**Comment by joshgreenwald** [comment]
> Am I missing something?

I have the same issue, but the code does compile and execute. Visual Studio thinks there is a problem but there seemingly isn't?

**Comment by almightyju** [comment]
> Am I missing something?

I assume VS code has the same issues as visual studio in that it sill shows the error, but if you build it should work. I just tried it on SDK 7.0.101 and it does run ok, have you tried `dotnet run` from the command line?

**Comment by almightyju** [comment]
> These will work in 7.0.0 or 7.0.1:
> <InputText @bind-Value="Value" @bind-Value:after="DoSomething" />
> <InputText @bind-Value:get="Value" @bind-Value:set="ValueChanged" />
> These will not work in either 7.0.0 or 7.0.1:
> <input @bind-value="Value" @bind-value:after="DoSomething" />
> <input @bind-value:get="Value" @bind-value:set="ValueChanged" />

These don't work in 7.0.0 without using rc2 nuget packages. I've just created a new project and changed the index code to this using this code:
```c#
@page "/"

<PageTitle>Index</PageTitle>

<input @bind-value="Value" @bind-value:after="ValueAfter" />

<EditForm Model="EditModel">
	<InputText @bind-Value="Value" @bind-Value:after="ValueAfter" />
</EditForm>

<ul>
	@foreach(string msg in Messages)
	{
		<li>@msg</li>
	}
</ul>

@code {
	string Value { get; set; } = string.Empty;

	private class Model
	{
		string Value { get; set; } = string.Empty;
	}
	Model EditModel { get; } = new();

	List<string> Messages { get; } = new();

	async Task ValueAfter()
	{
		Messages.Add(DateTime.Now.ToShortTimeString());
		await Task.Yield();
	}
}
```

and it does work altho visual studio shows it as an error it runs without issue, interestingly using 
```c#
<input @bind-value="Value" @bind-value:after="ValueAfter" @bind-value:event="oninput" />

<EditForm Model="EditModel">
	<InputText @bind-Value="Value" @bind-Value:after="ValueAfter" @bind-Value:event="oninput"/>
</EditForm>
```

doesnt for the InputText, I'm too tired to look into why tho.

Have you made sure to use the 7.0.1 packages? The default from a new project is
```xml
    <PackageReference Include="Microsoft.AspNetCore.Components.WebAssembly" Version="7.0.1" />
    <PackageReference Include="Microsoft.AspNetCore.Components.WebAssembly.DevServer" Version="7.0.1" PrivateAssets="all" />
```

**Comment by iphdav** [comment]
@almightyju You have actually highlighted why we all seem confused about if anything was fixed.  I was using *Blazor Server* where this is *not* fixed; and clearly some fixes have come through for blazor wasm.

Just to be clear: I have a clean laptop with .NET 7, just upgraded to 7.0.1 (SDK 7.0.101).  It has never had any release candidate bits and does not have any preview versions of anything.  Just a nice clean laptop, and thus I am reporting the issues most developers will face when looking at 7.0.1.  But my results were for Blazor Server.

**Comment by almightyju** [comment]
@iphdav Ahh, yea I'm only using wasm stuff so I haven't tried out the server side, this whole thing seems to be one heap of confusion left and right 🤷 

Turns out using the `<InputText ... @bind-Value:event="oninput" />` in wasm errors when changing values with 
```
rit: Microsoft.AspNetCore.Components.WebAssembly.Rendering.WebAssemblyRenderer[100]
      Unhandled exception rendering component: Object of type 'Microsoft.AspNetCore.Components.ChangeEventArgs' cannot be converted to type 'System.String'.
System.ArgumentException: Object of type 'Microsoft.AspNetCore.Components.ChangeEventArgs' cannot be converted to type 'System.String'.
   at System.RuntimeType.CheckValue(Object& value, ParameterCopyBackAction& copyBack, Binder binder, CultureInfo culture, BindingFlags invokeAttr)
   at System.Reflection.MethodBase.CheckArguments(Span`1 copyOfParameters, IntPtr* byrefParameters, Span`1 shouldCopyBack, ReadOnlySpan`1 parameters, RuntimeType[] sigTypes, Binder binder, CultureInfo culture, BindingFlags invokeAttr)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters)
   at System.Delegate.DynamicInvokeImpl(Object[] args)
   at System.MulticastDelegate.DynamicInvokeImpl(Object[] args)
   at System.Delegate.DynamicInvoke(Object[] args)
   at Microsoft.AspNetCore.Components.EventCallbackWorkItem.InvokeAsync[Object](MulticastDelegate delegate, Object arg)
   at Microsoft.AspNetCore.Components.EventCallbackWorkItem.InvokeAsync(Object arg)
   at Microsoft.AspNetCore.Components.ComponentBase.Microsoft.AspNetCore.Components.IHandleEvent.HandleEventAsync(EventCallbackWorkItem callback, Object arg)
   at Microsoft.AspNetCore.Components.EventCallback.InvokeAsync(Object arg)
   at Microsoft.AspNetCore.Components.RenderTree.Renderer.DispatchEventAsync(UInt64 eventHandlerId, EventFieldInfo fieldInfo, EventArgs eventArgs)
```

So I guess we might be a smidge closer to a working version 😅

**Comment by danroth27** [comment]
@iphdav When I try the originally reported repro steps I'm seeing that the app does successfully build and run, but I get errors in VS 😞:

![image](https://user-images.githubusercontent.com/1874516/207474305-3171f37d-6463-4b0e-8f78-52bd2e558e4a.png)

The app builds cleanly for me from the command-line:

![image](https://user-images.githubusercontent.com/1874516/207474602-c371e599-abd8-4c1a-be77-998c9648a3fa.png)

Is that consistent with what you are seeing? You indicated you were still seeing issues with building from the command-line?

**Comment by iphdav** [comment]
@danroth27 I also confirm the above bind:get and bind:set compiler error only happen with the input tag, and it works fine with InputText, InputNumber (etc).  Although don't forget to include the @using Microsoft.AspNetCore.Components.Forms if you want to try this.  You can maybe tell users to do this as a workaround until the team can address the bind:get and bind:set bugs.

**Comment by javiercn** [comment]
Hi folks, a few updates here.

The bits are on the 7.0.1 SDK, so you should be able to build/run/publish from the command line. Check the repo [here](https://github.com/javiercn/BindGetSetAfter701) for a working sample. 

There are a couple of things that might get in your way:
* Setup a global.json to ensure the right SDK version is used https://github.com/javiercn/BindGetSetAfter701/blob/main/global.json as you might also have 7.0.200 and the changes did not flow into that version.
* Clean your bin/obj folders to ensure that you are using up to date bits.
* If you are using Visual Studio, the compiler update is not there yet and you'll see squiggly errors, but your build should succeed despite that (if it does not, the command line build should work).
* Finally, using `[Parameter] public EventCallback<string> ValueChanged { get; set; }` is not valid as that is what caused the original issue with additional renders. This page contains a sample of all the valid combinations https://github.com/javiercn/BindGetSetAfter701/blob/main/BindGetSetAfter701/Client/Pages/Index.razor

**Comment by iphdav** [comment]
@javiercn Your repository is not public, could you please share it?

Your last point is implying the current documentation is wrong.  If that is the case please fix the documentation (or hopefully I can see your repo).  *So, this is incorrect now?*:
![image](https://user-images.githubusercontent.com/117321343/207589518-69a35835-bf75-4f2a-b35e-0565ae8c226c.png)

**Comment by javiercn** [comment]
@iphdav it should be public now.

@guardrex can you adjust the documentation?

**Comment by guardrex** [comment]
Yes, I'll take care of it this morning on the existing issue that @iphdav opened. I'll leave that issue open because I guesstimate that there will be developments on this in January. I'll keep an 👁️ out here for further developments.

**Comment by iphdav** [comment]
@javiercn Thank you so much for putting all those tests and examples together.  Yes I can access your repo now.

We just need all the documentation updated, and hopefully some community videos explaining that previous EventCallback example is no longer valid and all should be good.

As an aside...if you *don't* have this using in the page:
```razor
@using Microsoft.AspNetCore.Components.Forms
```
But you have an Input* control like this (or anything similar):
```razor
<InputText @bind-Value="_text" @bind-Value:after="() => { }" />
```

You get a very unhelpful error message:
```
The attribute names could not be inferred from bind attribute 'bind-Value'. Bind attributes should be of the form 'bind' or 'bind-value' along with their corresponding optional parameters like 'bind-value:event', 'bind:format' etc.	BlazorApp	C:\Users\dav\temp\blazortest701\BlazorApp\Pages\Index.razor	25	
```
The actual issue is you thought you were using a Razor component (InputNumber or InputText).  Obviously this is not a bug, and I realise Visual Studio should do it's highlighting thing and if the using is there it should appear in a different color.  But I actually suspect this will throw a lot of beginners.  It would be nice if there was something that could be done to improve the error - but it isn't obvious to me what that would be.

Thanks again.

**Comment by almightyju** [comment]
@javiercn should the following work?
```razor
<InputText @bind-Value="_text" @bind-Value:after="DoSomething" @bind-Value:event="oninput" />
```

Because it doesn't in web assembly at least, see my post above: https://github.com/dotnet/aspnetcore/issues/44957#issuecomment-1350068338. Not sure if it would need it's own issue tho if it is indeed broken

**Comment by javiercn** [comment]
@almightyju No. 

`@bind-Value:event` only applies to elements.

**Comment by guardrex** [comment]
@javiercn ... Seems like you meant to use the `Set` and `After` methods for the first examples of each (i.e., `SetAsync` and `AfterAsync` are duplicates) ...

https://github.com/javiercn/BindGetSetAfter701/blob/main/BindGetSetAfter701/Client/Pages/Index.razor

... and the following two lines are duplicated ...

```razor
<input type="text" @bind:get="_text" @bind:set="(value) => { }" />
```

```razor
<InputText @bind-Value:get="_text" @bind-Value:set="(value) => { }" />
```

Because I was in a major rush trying to get out of here this morning, I didn't catch these. I'm fixing them in the doc content now, which will go live in a few minutes.

**Comment by javiercn** [comment]
We think we had already addressed this issue at the CLI/SDK level, and we are working to get the fix into a future VS update. If you still have issues after updating to 7.0.1 and following the steps described [here](https://github.com/dotnet/aspnetcore/issues/44957#issuecomment-1351127371).

**Comment by paulguz-datapa** [comment]
@giarico 7.0.2 comes with Visual Studio 17.4.4, and the fix works for me.

**Comment by giarico** [comment]
I have updated to that version but i still have the same error.
cannot convert from 'Microsoft.AspNetCore.Components.EventCallback' to 'System.Action<string?>'

... even if it compiles

**Comment by GerardSmit** [comment]
> I have updated to that version but i still have the same error. cannot convert from 'Microsoft.AspNetCore.Components.EventCallback' to 'System.Action<string?>'
> 
> ... even if it compiles

@giarico This was already discussed in this issue https://github.com/dotnet/aspnetcore/issues/44957#issuecomment-1351127371:

> Finally, using `[Parameter] public EventCallback<string> ValueChanged { get; set; }` is not valid as that is what caused the original issue with additional renders. You need to pass in an action or a task returning function to set/after. This page contains a sample of all the valid combinations [javiercn/BindGetSetAfter701@main/BindGetSetAfter701/Client/Pages/Index.razor](https://github.com/javiercn/BindGetSetAfter701/blob/main/BindGetSetAfter701/Client/Pages/Index.razor?rgh-link-date=2022-12-14T11%3A45%3A43Z)

This is also documented in the docs (https://learn.microsoft.com/en-us/aspnet/core/blazor/components/data-binding):

![image](https://user-images.githubusercontent.com/2109929/212047619-d5754cc6-bbfb-44cd-b100-7d3db74d98d3.png)

It's documented how you can use actions here: <https://learn.microsoft.com/en-us/aspnet/core/blazor/components/data-binding#binding-with-component-parameters>

**Comment by giarico** [comment]
I'm using the feature in the right way and it compiles.
But i still have the visual error on visual studio.

**Comment by CrahunGit** [comment]
Me too. It compiles and runs fine but errors are showing all the time with latest visual studio

**Comment by danroth27** [comment]
@giarico @CrahunGit Can you confirm that you have .NET SDK version **7.0.102** installed? It comes with the recent VS update, which is **17.4.4**.

**Comment by giarico** [comment]
@danroth27 I confirm

**Comment by mbobka** [comment]
@danroth27 For me the problem occurs only in WebAssembly Blazor project. In Server Blazor project I have no problem.

**Comment by CrahunGit** [comment]
> @giarico @CrahunGit Can you confirm that you have .NET SDK version **7.0.102** installed? It comes with the recent VS update, which is **17.4.4**.

Yes. I'm using latest visual studio and 7.0.102 sdk version

**Comment by javiercn** [comment]
@giarico @mbobka @CrahunGit is the issue happening when you build in Visual Studio or also on the command line. Does it go away when you add a global.json file to the project/solution?

**Comment by giarico** [comment]
Hello, the issue is happening only in visual studio. No report reported when using the command line. Nothing change with global.json

**Comment by javiercn** [comment]
@giarico thanks for the additional details. I suspect that even if you see errors in VS, you should still see the Build Succeeded at the bottom of the IDE and should be able to run your app. Changing the error window from Build+Intellisense to Build might also be required.

@danroth27 this seems to be missing in the VS bits. The SDK doesn't matter here.

**Comment by giarico** [comment]
Yes, I see Build Succeded and the app runs correctly.
But hot reload doesn't work because for VS there is an error.
Every time I have to stop & restart the application. Really annoying.

**Comment by danroth27** [comment]
Hi folks. The VS fix for this just missed the window for 17.4.4 but should be addressed in the next VS patch update in February. We apologize for the wait and thank you for your patience!

**Comment by guardrex** [comment]
I'll leave our doc NOTE in place until the February release.

**Comment by Alerinos** [comment]
@guardrex Do we know the date mid february?

**Comment by guardrex** [comment]
I don't, but I would expect sometime this week.

**Comment by aaronsmithuk** [comment]
Previously the code below produced error: _cannot convert from 'Microsoft.AspNetCore.Components.EventCallback<int>' to 'System.Action<string?>'_

I no longer experience the build error in VS Community **17.4.5**

```
<p>@TestIt</p>
<select @bind="TestProp" @bind:after="TestBindAfter" class="form-select">
    <option value="1">Op 1</option>
    <option value="2">Op 2</option>
    <option value="3">Op 3</option>
    <option value="4">Op 4</option>
</select>
```

```
@code {
    public int TestProp { get; set; } = 1;
    private string TestIt = "";
    private void TestBindAfter()
    {
        TestIt = "Chosen option = " + TestProp.ToString();
    }
}
```

**Comment by guardrex** [comment]
It doesn't seem fixed in VS Community 17.5.0 Preview 6.0 ...

![image](https://user-images.githubusercontent.com/1622880/219396638-54d65382-0969-447b-9d52-21fb325c9bb1.png)

**Comment by paulpitchford** [comment]
It's still not working on a child component for me in Community 2022 x64 17.4.5. Had hopes from what I'd read that this would be rectified.

![image](https://user-images.githubusercontent.com/13196888/219950926-6ede00dc-375a-4407-84c3-02deee46325e.png)

**Comment by puppetSpace** [comment]
> It doesn't seem fixed in VS Community 17.5.0 Preview 6.0 ...
> 
> ![image](https://user-images.githubusercontent.com/1622880/219396638-54d65382-0969-447b-9d52-21fb325c9bb1.png)

Don't think the preview editions have this update already. I also have 17.5 preview 6 and it doesn't work. Also don't see anything in the release notes for this [Preview](https://learn.microsoft.com/en-us/visualstudio/releases/2022/release-notes-preview#17.5.0-pre.6.0)

**Comment by JvanderStad** [comment]
I just have installed Visual Studio Professional 17.5.0. I did not have an error before, but now I do.

After installation I have the same error: Both in Visual Studio and with `dotnet build`

`error CS1503: Argument 3: cannot convert from 'Microsoft.AspNetCore.Components.EventCallback<string>' to 'System.Action<string?>'`

<img width="739" alt="image" src="https://user-images.githubusercontent.com/5426370/220507281-7538973c-d63b-41ea-8034-7b6bc88915d8.png">

```
.NET SDK:
 Version:   7.0.200
 Commit:    534117727b

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.22621
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\7.0.200\

Host:
  Version:      7.0.3
  Architecture: x64
  Commit:       0a2bda10e8
```

When I add a `global.json` with a previous sdk it gives no error
```json
{
  "sdk": {
    "version": "7.0.103"
  }
}
```

**Comment by Jarek300** [comment]
The same error in 17.5.0 RTM.

**Comment by samueldjack** [comment]
I have the same scenario as JvanderStad.

I first noticed a problem on our build server last Friday, which suddenly started failing because it automatically rolled forward the SDK from 7.0.103 (which worked) to 7.0.200 which failed with the "Cannot convert" errors.

So I installed 7.0.200 locally (VS 17.5 hadn't been released then), and I started seeing the error locally.

I've just tried installing 17.5 and I still have the problem.

The only workaround at the moment is using global.json with an sdk version of 7.0.103

**Comment by guardrex** [comment]
I just reached 17.6 Preview 1 with VS Preview. The problem persists.

**Comment by crmeadowcroft** [comment]
It also was fixed for me in 17.4 but broken again in 17.5

**Comment by iphdav** [comment]
@bind:after has been broken again in V17.5.0.  It is not just intellisense this time - the actual build fails.  We are now tracking it here:
https://github.com/dotnet/razor/issues/8315

**Comment by jaredpar** [comment]
Sorry, this is indeed broken in 17.5 / 7.0.200. 

The reason for this break is that between 17.4 and 17.5 the razor compiler code based moved from [dotnet/razor-compiler](http://github.com/dotnet/razor-compiler) to [dotnet/razor](http://github.com/dotnet/razor). This move was done so the razor compiler and IDE / tooling experience could exist in a single repository. That will allow us to move together as a team, make cross cutting changes faster and overall improve the razor experience. It's an approach we've had great success with in roslyn. 

Unfortunately the fix for the `@bind:after` scenarios was merged into razor-compiler at the exact time we were doing the migration between the repositories. Our team got quickly consumed by cleaning up the result of merging the repositories and missed that this change was not correctly ported over. As soon as we saw this bug we realized our mistake and are working to correct it.

I'm sorry for the disruption this has caused. We'll be patching 17.5 and 7.0.2xx as soon as we can.

**Comment by FLAMESpl** [comment]
@bind:set also do not work

**Comment by dahln** [comment]
Editing my WASM project that uses bind:after, and got this error:

"cannot convert from 'Microsoft.AspNetCore.Components.EventCallback<string>' to 'System.Action<string?>"

running VS 17.5.1 and SDK 7.0.201.

Get the same error if I run 'dotnet build' from the terminal

**Comment by killnine** [comment]
Watching this issue. Just started a Blazor project the other week and thought I was going crazy not being able to resolve the issue in my markup. Eagerly awaiting a fix. Thanks, team, for your hard work.

**Comment by crmeadowcroft** [comment]
Seems to be fixed with today's update

**Comment by guardrex** [comment]
@danroth27 ... I'm seeing the same as @crmeadowcroft ... 17.6 Preview 2 seems to have licked it here. Shall I proceed to remove the documentation notes that we're keeping on the VS UI problems?

**Comment by Alerinos** [comment]
@guardrex This scenario still doesn't work, or did I miss something and it changed?
```razor
<input type="datetime-local" @bind="data.Date" @bind:after="(args) => { CheckedChanged(data); }"  min="2000-06-07T00:00" max="@DateTime.UtcNow.ToString("o")" />
```

**Comment by crmeadowcroft** [comment]
@Alerinos, I think your issue is that @bind:after doesn't have any parameters. You can get the most recent value from the variable you @bind to if you need it in this method.

**Comment by Alerinos** [comment]
@crmeadowcroft Unfortunately, this is my only solution. I have a table of multiple dates, when I change it a data calculation is created. I need to pass the model or record ID. This is currently the only way.

**Comment by danroth27** [comment]
> @danroth27 ... I'm seeing the same as @crmeadowcroft ... 17.6 Preview 2 seems to have licked it here. Shall I proceed to remove the documentation notes that we're keeping on the VS UI problems?

Yup, we should be ok to remove the notes at this point.

**Comment by danroth27** [comment]
> @guardrex This scenario still doesn't work, or did I miss something and it changed?
> 
> ```
> <input type="datetime-local" @bind="data.Date" @bind:after="(args) => { CheckedChanged(data); }"  min="2000-06-07T00:00" max="@DateTime.UtcNow.ToString("o")" />
> ```

Looks like you're hitting a limitation of the feature. Do you want to open a new issue to propose support for taking the bound value as a parameter?

**Comment by crmeadowcroft** [comment]
@Alerinos couldn't you just write @@bind:after="() => { CheckedChanged(data); }"

**Comment by javiercn** [comment]
@danroth27 passing the argument to bind:set is the difference between set and after, so adding that would make things confusing, hence we would not do it.

**Comment by MisinformedDNA** [comment]
> @danroth27 passing the argument to bind:set is the difference between set and after, so adding that would make things confusing, hence we would not do it.

The example shown doesn't pass an argument to `set`, it passes it to `after`. Or am I missing something?

**Comment by javiercn** [comment]
@MisinformedDNA after does not receive an argument, it is already available in the property/field you bound to. If you want access to the argument (before it is set) you use `bind:set`.

**Comment by Westboldyi** [comment]
Is it still the case that when using `bind:set` that the new value must be assigned before any async work is allowed? referring to the design proposal: https://github.com/dotnet/aspnetcore/issues/39837

**Comment by javiercn** [comment]
Yes, otherwise renders might happen with the incorrect value.

**Comment by ladeak** [comment]
For me this is looks still broken in 17.6 Preview 1, as well as when I do dotnet build.

```
.NET SDK:
 Version:   8.0.100-preview.1.23115.2
 Commit:    913e4679b3
```

Could you add `Warning` or a `Notice` to the documentation https://learn.microsoft.com/en-us/aspnet/core/blazor/components/data-binding?view=aspnetcore-7.0 here, it is really misleading now?

**Comment by guardrex** [comment]
@ladeak ... Try 17.6 Preview 2. We just pulled the notice from the docs with Dan's blessing. He'll let me know if something comes up and we need to revert back to the notice.

**Comment by danroth27** [comment]
> Sorry, that is my mistake. My VS is actually 17.6 Preview 2.

Could you please also try with .NET 8 Preview 2 and see if that solve the problem for you?

**Comment by ladeak** [comment]
After installing SDK version 8.0.0-preview.2.23128.3 both the CLI and the VS works as documented.
